### ISSUE TITLE
Minimize to tray - "Run in Background" feature

### ISSUE THREAD

**Issue by Stamimail** [foudfou/FireTray#164]
The feature means: when starting the browser, the browser is opend and there is a system tray icon.
This single icon is always there, even when closing all windows. Each clicking on the icon opens a new window (Ctrl+N).

https://bugzilla.mozilla.org/show_bug.cgi?id=932772

**Comment by foudfou** [comment]
Hi, this is a feature request right ?
I think FireTray has already similar functionalities with:
* *Start hidden to tray*, see the FireTray preferences
* click on the icon to show/hide all windows
* right-click on the icon to open a menu with a *New window* item
So I'm tempted to close this one as *invalid*.

Also seems related to #144 in that restarting FF/TB is too lengthy.

**Comment by Stamimail** [comment]
Yes, this is a feature request.
I checked FireTray, and it hasn't.

Start hidden to tray needs to be disabled.
click on the icon to show/hide all windows also needs to be diabled.

I think this feature was well explained. If not, please tell me.

**Comment by foudfou** [comment]
Oh I see: you'd like FF to start and display normally, but when closing *all* windows, it somehow stays open.
I think this is very similar to #144. If you agree, I'd like to close this one as a duplicate, and discuss further on #144. More specifically, have you tried the *Only last window can be hidden* [other windows will be closed (terminated)] preference ?

**Comment by Stamimail** [comment]
>"you'd like FF to start and display normally"
Yes. And when FF start, there will be a system tray icon too. Clicking on this icon, open new windows.
>"but when closing all windows, it somehow stays open."
No matter how many windows open,  and no matter if all the windows closed, This single System Tray Icon (FF icon) is always there.

Do you think it needs to create a new FF extension for "Run in Background" (How difficult would it be?), or it's related to FireTray?

>"More specifically, have you tried the Only last window can be hidden..."
I didn't understand.

![firetray preference](https://cloud.githubusercontent.com/assets/11010156/6205150/28d3b6bc-b56d-11e4-8586-18bf5c5a03c1.PNG)

**Comment by foudfou** [comment]
> Do you think it needs to create a new FF extension for "Run in Background" (How difficult would it be?), or it's related to FireTray?

My understanding is that FF is a basically GUI application. I don't know any headless mode. So "Run in Background" doesn't really make sense. The closest thing you can get is probably hiding windows. FireTray can do that.
I think the crucial point is: what happens when you close the last window ? Or possibly the last tab ?
Obviously guys like you and @informatorius would like the last window to disappear, or maybe have it invisible, remaining on the default homepage.

If you really want exactly what you described, I guess you should fork FireTray,  keep only the necessary parts, and work from there.

My mistake: *Only last window can be hidden* was not available under Windows. I just fixed that (0f79fe2). [Here's an xpi](https://cloud.foudil.fr/public.php?service=files&t=770c7f4d4862e738226c9ffc351cf036).

**Comment by Stamimail** [comment]
I can't give you technically answer, I'm not a developer (I can't fork too). But I can say, that what is expected is opening a FF window in Homepage.

**Comment by foudfou** [comment]
@Stamimail have you tested the *Only last window can be hidden* preference [described above](#issuecomment-74442507) ?

**Comment by informatorius** [comment]
I understand Stamimail this way:

1) Autostart Firefox to Tray
Current behavior: If configured in FireTray "Start Application hidden to tray" and firefox is included in Windows Startup folder => it works
Expected behavior: Have an FireTray options "Start with Windows hidden to tray"
Todo: Make the FireTray option "Start with Windows hidden to tray" which also adds firefox to Windows Startup folder if not already exist

2) Click on FireTray icon
Current behavior: Click on FireTray icon hides/shows all windows.
Expected behavior: Click on FireTray opens new window with Homepage.
Todo: Make FireTray option that enables this behavior change.

3) Click on Firefox close X button
Current behavior: Window is minimized to tray or closed, depending on FireTray option
Expected behavior: Close to tray, i.e. window is closed, but last window/tab is set to homepage and minimized to tray.
Todo: Make FireTray option CloseToTray and when enabled: 
a) CloseToTray enabled: on window close X button click closes the window if there is another window open
b) CloseToTray enabled: If it is a single window then close all tabs and in last tab load homepage and minimize window to tray
c) CloseToTray enabled: If firefox is configured for warnonclosetabs then this firefox dialog shows. If dialog is cancelled then nothing happens. If dialog is confirmed then a) or b).

This is similar functionality i requested in issue #144
I think these options can be included in FireTray without high effort.
But it can also be seen as unique "FireCloseToTray" functionality to fork FireTray to a new addon and strip all the minimize functionality and options, while FireTray stays with the "FireMinimizeToTray" functionality.

The Master will decide.

**Comment by informatorius** [comment]
I tried xpi, option "close on last tab only" is available now on Windows 7 Firefox 35 :-)

**Comment by Stamimail** [comment]
@foudfou , When I clicking on "Here's an xpi." FF opens blank page with:
"This connection is not safe"

**Comment by foudfou** [comment]
My server has a self-signed certificate. Just bypass the warning.

**Comment by Stamimail** [comment]
It's not exactly the same.
Please take a look in this:
http://www.ge.tt/5RtdUfA2/v/0?c

This is Everything Search Portable, with my desired settings.

**Comment by Stamimail** [comment]
1) Autostart Firefox to Tray
I think both is needed:
"Start Application hidden to tray" & "Start with Windows hidden to tray"
I for example would prefer the first disable and the second enabled.
Now I can see that in Everything, the second is achieved by a command line Parameter:
"C:\Portable Program Files\Everything-1.3.4.686.x64\Everything.exe" -startup

2) Click on FireTray icon
Yes. Click on FireTray (FF tray icon) opens new window with Homepage.

3) Click on Firefox close X button
I'm not sure the name CloseToTray is the right name, because other apps can use this term for each window, so you have multiple icons in the system tray.

BTW, can you fix the Preference Dialog Box Height?

**Comment by Stamimail** [comment]
Do you think FireTray can't let the users to choose what  will happen when clicking on the Tray Icon:
1. Minimize to tray / UnMinimize (Current)
or
2. Open New HomePage window.
?

**Comment by foudfou** [comment]
@Stamimail open window when clicking on the icon could be an option. Although I'm not sure what it's supposed to do under Thunderbird or other applications (Zotero?). I have the overall feeling this would bring unnecessary complexity.
But the main reason for not working on this now are: first I don't see it as a priority compared to other issues, and second I don't know if it's a feature a lot of users would want.
So let's keep your request open and see if others show some interest.

**Comment by Stamimail** [comment]
Right now I'm using what you brought me, and I have to say that although it is not ready, it is very helpful what it prevents the application  from exit. Thank you.
I don't think this topic will get a lot of views, so you could have replies about this idea.
In the meantime, if you have any idea for me I would happy to hear. Thank you very much.

**Comment by Stamimail** [comment]
When I watch a video in the browser, and want to close this single window, I can hear the sound of the video after closing it. It seems that simple hiding is not the required solution.

**Comment by foudfou** [comment]
When *closing window hides to tray* is set, you should close windows with Ctrl-W.

**Comment by Stamimail** [comment]
My settings:
![firetray2](https://cloud.githubusercontent.com/assets/11010156/6451143/24291390-c135-11e4-956d-c6a93ae2b7b1.PNG)

The problem: Ctrl+W exit the program.
Expected: Ctrl+W should be the same like X closa of the window or Alt+F4.

Perhaps, when closing the last window, make the window blank. After when clicking the System Tray icon will open in HomePage.
